summariser/src/scenario/validation_receipts.rs

```diff
@@ -0,0 +1,51 @@
+use crate::analyze::{partitioned_rate_stats, partitioned_timing_stats};
+use crate::model::{PartitionedRateStats, PartitionedTimingStats, SummaryOutput};
+use crate::query;
+use crate::query::zome_call_error_count;
+use anyhow::Context;
+use serde::{Deserialize, Serialize};
+use wind_tunnel_summary_model::RunSummary;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct ValidationReceiptsSummary {
+    receipts_complete_timing: PartitionedTimingStats,
+    receipts_complete_rate: PartitionedRateStats,
+    error_count: usize,
+}
+
+pub(crate) async fn summarize_validation_receipts(
+    client: influxdb::Client,
+    summary: RunSummary,
+) -> anyhow::Result<SummaryOutput> {
+    assert_eq!(summary.scenario_name, "validation_receipts");
+
+    let receipts_complete = query::query_custom_data(
+        client.clone(),
+        &summary,
+        "wt.custom.validation_receipts_complete_time",
+        &["agent", "op_type"],
+    )
+    .await
+    .context("Load receipts complete data")?;
+
+    SummaryOutput::new(
+        summary.clone(),
+        ValidationReceiptsSummary {
+            receipts_complete_timing: partitioned_timing_stats(
+                receipts_complete.clone(),
+                "value",
+                "10s",
+                &["agent", "op_type"],
+            )
+            .context("Receipts complete timing")?,
+            receipts_complete_rate: partitioned_rate_stats(
+                receipts_complete.clone(),
+                "value",
+                "10s",
+                &["agent", "op_type"],
+            )
+            .context("Receipts complete rate")?,
+            error_count: zome_call_error_count(client, &summary).await?,
+        },
```

Comment:
I don't love how you have to manually write one of these for every scenario. I totally see you trying to make this as declarative as possible to remove friction/boilerplate, and we don't have to get all the way there on the first try... but some questions for how you see this evolving in the future:

- Do you think there's a way to specify this closer to the scenario? I.e. if part of the scenario trait was to define a handler function for how to generate the summary, so we'd not have to cross crates when writing scenarios?
- Is there a way to go all the way declarative, like by specifying whether it should be partitioned rate stats or timing stats somehow in the field itself and having the summariser parse the field names to know if and how it should summarize them?

Reply:
Yep, objections understood. I'm not that keen to have the analysis logic in the scenarios because they're meant to get uploaded/run on target nodes that aren't responsible for analyzing. So I'm picturing those being two separate steps and so it made sense to me to separate the code out. Because they're separate though, I think it would be valid to do one PR that adds a scenario and we can experiment with the scenario for a while and come back and add analysis when we're ready. The tool will just ignore scenarios it doesn't know how to create a summary for.

As for whether we can make this declarative? Maybe, I'm thinking that what I've done is the minimum and some scenarios are going to need to do more complicated things and examine a mixture of automatically captured "client" metrics along with custom metrics. I think we'll probably limit ourselves by going down that route, or end up with a complex DSL of some kind for asking what metrics we want. Or I'm wrong and this is as complicated as the analysis needs to get and maybe we'll see we can go that route in the near future. 

Basically, I had a reason for separating the code out from the scenarios but what we need next is probably going to depend on what tools we pick and how useful we find the data and whether we need more complex insights or get plenty of value from these few analysis functions... I really don't know :) 

Something I do have planned that would possible be more complicated if we put analysis information in the scenarios is creating a change summary between two reports. That would involve compiling all the scenarios together if the equivalent of these models lived in each scenario. Having them in this crate should make it easier to define a "difference" trait of some kind that can use these models. But then again, if it was all declarative maybe we wouldn't need custom models and could just use a field to type the various analysis outputs and run a standard diff over them... Maybe! :)

One way or another, I'm certainly open to this being evolved!

Reply:
> come back and add analysis when we're ready

:+1:

> I think we'll probably limit ourselves by going down that route

I agree we'll probably want more specialized analysis for specific problems that come up. I almost wish I had never used splunk... It would be trivial to pick two or more run ids out of the data, then pick the first timestamp for each run id and calculate a field that is just time-since-start so we can overlay and manipulate the graphs however we want to get at the visualization or summary we want.

Unfortunately, what I've seen so far experimenting with openobserve, it is not that easy to manipulate.

---

Okay, cool. Things to keep thinking about as we move forward.